### ISSUE TITLE
possible to reintroduce type hints for php and/or type hint in comments?

### ISSUE THREAD

**Issue by GreyCat** [kaitai-io/kaitai_struct#399]
Original issue was raised by @coniurare in https://github.com/kaitai-io/kaitai_struct_compiler/issues/136:

> Hi,
> 
> thanks for such a nice library and all the work that goes into it. @GreyCat @markbook2 I remember that the type hints for the php runtime were removed due to problems with nullable types? How hard would it be to add them back now that php 7.1 and 7.2 are out with type hints like `?Foo` - maybe behind a command line switch for example `--add-php-type-hints` or by using two php runtime targets on the command line e.g `--target php` (no type hints) and `--target php:7.1+` (has type hints)?  
> 
> Also, how hard would it be to type hint arrays of objects by using doc comments? E.g
> 
> ```php
> ...
> /*
>  * @return Foo[]
>  */
> public function foos(): array {
>     ...
> }
> ...
> ```
> This would help IDEs to autocomplete for example in foreach etc. The second one is possible by including it in the `doc:` part of the .ksy file but a automatic solution would be nice.

**Comment by GreyCat** [comment]
So, there are separate things here:

### Addition of type enforcement annotations using php7.1+ nullable types syntax

The main question is whether it will require separate set of test specs or not.

* If no, then it's relatively easy. CI integration would be something close to our Python 2 vs Python 3 testing, but we'll need to produce two sets of PHP target files - i.e. for "legacy PHP" and "for PHP 7.1+".
* If yes, then it's relatively complex and time-consuming: we'll need to build a new set of specs and support it. Several months ago it would require *a lot* of manual effort, right now it's easier thanks to KST (https://github.com/kaitai-io/kaitai_struct/issues/253), but (1) KST spec autogeneration still covers only ~70% of tests, (2) it's still some work to introduce that 

### Addition of `@return` in docstrings

Relatively easy, as it requires only manual testing and backwards compatible. Would you want to:

* add it only if docstrings are present, or generate it even if docstrings are not present?
* do it for all types (i.e. including native types), or only for user types?
* always output `@return Foo`, or only for arrays - i.e. `@return Foo[]`?

**Comment by coniurare** [comment]
> Addition of type enforcement annotations using php7.1+ nullable types syntax
> - The main question is whether it will require separate set of test specs or not.

I don't know, only started playing around with kaitai recently after looking for something like the templates from 010 editor.

> Addition of `@return` in docstrings
> - add it only if docstrings are present, or generate it even if docstrings are not present?

Personally I would prefer if it's generated even if no docstring is present, but I may be obsessive with that since I generally annotate every member (e.g `@var int`) and method (`@param` and `@return`).  Even in code that is not a public api.

> - do it for all types (i.e. including native types), or only for user types?

If type hints for 7.1+ would come back it would be redundant I guess. Otherwise I would prefer it for all types. Again, I might be obsessive with that coming from a static language and wanting to know all my types.

> - always output `@return Foo`, or only for arrays - i.e. `@return Foo[]`?

I would prefer always using `@return Foo` and not only for the collection/array cases. If the returned type has many members/methods it's easy for a IDE to autocomplete them. But if type hinting comes back it would only be helpful in the array case.

At the moment I use a helper function to help the IDE e.g

```php
<?php declare(strict_types=1);

use Kaitai\Struct\Struct;
use Some\Namespace\SomeFileFormat;

function ideHelper(Struct $struct): SomeFileFormat {
    return $struct;
}

$data = SomeFileFormat::fromFile('/path/to/file');
// at this point my IDE (PhPStorm) can only autocomplete:
//    $data->_io();
//    $data->_parent();
//    $data->_root();

$data = ideHelper($data);
// at this point the IDE can autocomplete all the above and:
//    $data->someFieldInTheBinary();
//    $data->anotherFieldInTheBinary();
//    $data->...
```

**Comment by markbook2** [comment]
@GreyCat 

> The main question is whether it will require separate set of test specs or not.

The separate set of test specs is not necessary here.

Syntax of return types for PHP 7.1 is not back compatible for PHP 7.0. I propose to switch to 7.1 as it is [actively supported](http://php.net/supported-versions.php) and in general use only actively supported version of PHP. If someone need to use older version of PHP (< 7.1) then it can get respective code with older version of Kaitai Struct.

Additionally I propose to add the strict_types to catch more possible errors.
```
<?php declare(strict_types=1);
namespace Kaitai\Struct\Struct;
// ...
```

**Comment by markbook2** [comment]
The adding of doc comments like:
```php
...
/**
 * @return Foo[]
 */
public function foos(): array {
    ...
}
...
```
can be useful as it specifies that the array is array of Foo. Example of such `return` doc annotation can be found in [PHP-Parser](https://github.com/nikic/PHP-Parser/blob/249953472902fffa0b0f0ccbb4d6961859584caa/lib/PhpParser/NodeTraverser.php#L74).

**Comment by coniurare** [comment]
I fully support @markbook2's proposal of switching to PHP 7.1. Sounds like the most maintainable way.

**Comment by markbook2** [comment]
@GreyCat I don't think that there are many users that use only PHP >= 7.0 and < 7.1. But if you think that the PHP 7.0 should be supported it can be done.

Most compatible and easy to implement way is just adding `@return Foo[]` doc comments + adding of the mentioned above `declare(strict_types=1);` another option is support 2 implementations - one for PHP 7.0 and one for PHP 7.1. But in this case what if we will need to use features of PHP 7.2 like [object type](https://wiki.php.net/rfc/object-typehint) - support all these cases does not make sense. The most compatible way is using main not outdated version of PHP - which has [active support status](http://php.net/supported-versions.php) - currently it is 7.1 version, the 5.6 and 7.0 are outdated versions.

**Comment by GreyCat** [comment]
Ok, just to be clear here: I reckon that there are *many* older PHP installs in the wild, like PHP5:

* Ubuntu 14.04 Trusty which is still a supported LTS and lots of things are still based on that one sports php 5.5.9
* w3cschools has tutorials on "PHP 5" — https://www.w3schools.com/php/default.asp — this is exactly what first comes by googling for "php tutorial"
* Even Travis's default install is php 5.6.32

Right now we work with all of them. I don't see why we should break that compatibility, especially when the effort to support both older and newer PHP is relatively small.

**Comment by markbook2** [comment]
> Right now we work with all of them. 

Have you tried run a sample which uses KS PHP runtime under PHP < 7.x? Does it work?

As I see that we use PHP 7.1.11 on Travis.

As example of the current state of PHP some notable projects can be taken with theirs requirements:
- [Symfony 4 requires PHP >= 7.1.3](https://symfony.com/doc/current/reference/requirements.html)
- [PHPUnit requires PHP >= 7.1](https://phpunit.de/manual/current/en/installation.html)
- [Laravel requires PHP >= 7.1.3](https://laravel.com/docs/5.6/installation)
- [Yii 2.0 requires PHP >= 5.4](https://github.com/yiisoft/yii2-framework/blob/master/composer.json#L66) but will switch to PHP 7.1 for Yii 2.1
- [ZF requires PHP 5.6](https://docs.zendframework.com/tutorials/getting-started/skeleton-application/)

So some frameworks use PHP 5.x and some will switch to PHP 7.1. I don't see any framework who supports PHP >= 7.0, all of them have minimal requirement of PHP >= 7.1

**Comment by coniurare** [comment]
So, there seem to be 3 possible solutions? Not counting doing nothing, hehe:

- dropping support for PHP versions prior to 7.1 while adding `declare(strict_types=1)`, type hints (`function(): array`, `function(): ?CustomType` etc.) and on top of that adding a docblock in case a array is returned `@return CustomType[]`. Everything together making the code safer while also helping IDEs autocomplete and being relatively low effort. 

- still supporting PHP 5.6 which - like PHP 7.0 - is dead after December 2018 (no security updates anymore), but adding docblocks (`@return CustomType[]`, `@return CustomType`, `@return int` etc.) to all functions to help IDEs autocomplete the code, while also implementing a command line switch for the features outlined above for PHP 7.1 and above. Seems like the most effort.

- still supporting PHP 5.6 with the features outlined above, but _without_ a switch for users of PHP 7.1 and above. Then dropping support for PHP 5.6 after it's EOL in December 2018 and switching to only supporting PHP 7.1 and above. Seems like the least drastic change while also being relatively low effort.
